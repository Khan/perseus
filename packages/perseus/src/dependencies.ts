import * as React from "react";

import type {PerseusDependencies, PerseusDependenciesV2} from "./types";

let _dependencies: PerseusDependencies | null | undefined = null;

export const setDependencies = (dependencies: PerseusDependencies) => {
    _dependencies = dependencies;
};

export const getDependencies = (): PerseusDependencies => {
    if (_dependencies) {
        return _dependencies;
    }

    throw new Error(
        [
            "Perseus has not been provided required dependencies.",
            "setDependencies(dependencies) must be called first.",
            "Make sure Perseus is being imported from javascript/perseus/perseus.js.",
        ].join("\n"),
    );
};

export type DependencyProps = Partial<
    React.ContextType<typeof DependenciesContext>
>;

export const DependenciesContext = React.createContext<PerseusDependenciesV2>({
    analytics: {onAnalyticsEvent: async () => {}},
});

/**
 * useDependencies provides easy access to the current Perseus dependencies.
 */
export const useDependencies = (): PerseusDependenciesV2 => {
    const deps = React.useContext(DependenciesContext);

    return deps;
};
