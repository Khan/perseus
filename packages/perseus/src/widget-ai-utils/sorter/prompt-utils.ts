import type {WidgetType} from "../../prompt-types";
import type {PerseusSorterUserInput} from "../../validation.types";

export type SorterPromptJSON = {
    type: WidgetType;
    userInput: {
        values: PerseusSorterUserInput["options"];
        changed: PerseusSorterUserInput["changed"];
    };
};

export const getPromptJSON = (
    userInput: PerseusSorterUserInput,
): SorterPromptJSON => {
    return {
        type: "sorter",
        userInput: {
            values: userInput.options,
            changed: userInput.changed,
        },
    };
};
