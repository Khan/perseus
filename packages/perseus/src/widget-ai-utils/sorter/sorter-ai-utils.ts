import type {PerseusSorterUserInput} from "@khanacademy/perseus-score";

export type SorterPromptJSON = {
    type: "sorter";
    userInput: {
        values: ReadonlyArray<string>;
        changed: boolean;
    };
};

export const getPromptJSON = (
    userInput: PerseusSorterUserInput,
): SorterPromptJSON => {
    return {
        type: "sorter",
        userInput: {
            values: userInput.options,
            changed: userInput.changed,
        },
    };
};
