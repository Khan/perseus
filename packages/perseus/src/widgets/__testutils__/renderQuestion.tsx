import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
