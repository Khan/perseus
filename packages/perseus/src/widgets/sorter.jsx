// @flow
import PropTypes from "prop-types";
import * as React from "react";
import _ from "underscore";

import Sortable from "../components/sortable.jsx";
import {linterContextProps, linterContextDefault} from "../gorgon/proptypes.js";
import {ApiOptions} from "../perseus-api.jsx";
import Util from "../util.js";

import type {SortableOption} from "../components/sortable.jsx";
import type {WidgetExports} from "../types.js";

const {shuffle} = Util;
const HORIZONTAL = "horizontal";
const VERTICAL = "vertical";

class Sorter extends React.Component<$FlowFixMe, $FlowFixMe> {
    _isMounted: boolean;

    static propTypes = {
        apiOptions: ApiOptions.propTypes,
        // eslint-disable-next-line react/forbid-prop-types
        correct: PropTypes.array,
        layout: PropTypes.oneOf([HORIZONTAL, VERTICAL]),
        onChange: PropTypes.func,
        padding: PropTypes.bool,
        problemNum: PropTypes.number,
        trackInteraction: PropTypes.func.isRequired,
        linterContext: linterContextProps,
    };

    static defaultProps: $FlowFixMe = {
        correct: [],
        layout: HORIZONTAL,
        padding: true,
        problemNum: 0,
        onChange: function () {},
        linterContext: linterContextDefault,
    };

    static validate(state: $FlowFixMe, rubric: $FlowFixMe): $FlowFixMe {
        const correct = _.isEqual(state.options, rubric.correct);

        return {
            type: "points",
            earned: correct ? 1 : 0,
            total: 1,
            message: null,
        };
    }

    state: $FlowFixMe = {
        changed: false,
    };

    componentDidMount() {
        this._isMounted = true;
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    handleChange: (SyntheticInputEvent<>) => void = (e) => {
        if (!this._isMounted) {
            return;
        }

        this.setState({changed: true}, () => {
            // Wait until all components have rendered. In React 16, the
            // setState callback fires immediately after componentDidUpdate,
            // and there is no guarantee that parent/siblings components have
            // finished rendering.
            // TODO(jeff, CP-3128): Use Wonder Blocks Timing API
            // eslint-disable-next-line no-restricted-syntax
            setTimeout(() => {
                this.props.onChange(e);
                this.props.trackInteraction();
            }, 0);
        });
    };

    getUserInput: () => {|options: $FlowFixMe|} = () => {
        // eslint-disable-next-line react/no-string-refs
        return {options: this.refs.sortable.getOptions()};
    };

    moveOptionToIndex: (option: SortableOption, index: number) => void = (
        option,
        index,
    ) => {
        // eslint-disable-next-line react/no-string-refs
        this.refs.sortable.moveOptionToIndex(option, index);
    };

    simpleValidate: ($FlowFixMe) => $FlowFixMe = (rubric) => {
        // If this widget hasn't been changed yet, we treat it as "empty" which
        // prevents the "Check" button from becoming active. We want the user
        // to make a change before trying to move forward. This makes an
        // assumption that the initial order isn't the correct order! However,
        // this should be rare if it happens, and interacting with the list
        // will enable the button, so they won't be locked out of progressing.
        if (!this.state.changed) {
            return {
                type: "invalid",
                message: null,
            };
        }

        return Sorter.validate(this.getUserInput(), rubric);
    };

    render(): React.Node {
        const options = shuffle(
            this.props.correct,
            this.props.problemNum,
            /* ensurePermuted */ true,
        );

        const {apiOptions} = this.props;
        const marginPx = apiOptions.isMobile ? 8 : 5;

        return (
            <div className="perseus-widget-sorter perseus-clearfix">
                <Sortable
                    options={options}
                    layout={this.props.layout}
                    margin={marginPx}
                    padding={this.props.padding}
                    onChange={this.handleChange}
                    linterContext={this.props.linterContext}
                    // eslint-disable-next-line react/no-string-refs
                    ref="sortable"
                />
            </div>
        );
    }
}

export default ({
    name: "sorter",
    displayName: "Sorter",
    widget: Sorter,
    isLintable: true,
}: WidgetExports<typeof Sorter>);
