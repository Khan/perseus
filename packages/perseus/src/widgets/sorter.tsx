/* eslint-disable @khanacademy/ts-no-error-suppressions */
import {
    linterContextProps,
    linterContextDefault,
} from "@khanacademy/perseus-linter";
import PropTypes from "prop-types";
import * as React from "react";
import _ from "underscore";

import Sortable from "../components/sortable";
import {ApiOptions} from "../perseus-api";
import Util from "../util";

import type {SortableOption} from "../components/sortable";
import type {WidgetExports} from "../types";

const {shuffle} = Util;
const HORIZONTAL = "horizontal";
const VERTICAL = "vertical";

class Sorter extends React.Component<any, any> {
    // @ts-expect-error - TS2564 - Property '_isMounted' has no initializer and is not definitely assigned in the constructor.
    _isMounted: boolean;

    static propTypes = {
        apiOptions: ApiOptions.propTypes,
        // eslint-disable-next-line react/forbid-prop-types
        correct: PropTypes.array,
        layout: PropTypes.oneOf([HORIZONTAL, VERTICAL]),
        onChange: PropTypes.func,
        padding: PropTypes.bool,
        problemNum: PropTypes.number,
        trackInteraction: PropTypes.func.isRequired,
        linterContext: linterContextProps,
    };

    static defaultProps: any = {
        correct: [],
        layout: HORIZONTAL,
        padding: true,
        problemNum: 0,
        onChange: function () {},
        linterContext: linterContextDefault,
    };

    static validate(state: any, rubric: any): any {
        const correct = _.isEqual(state.options, rubric.correct);

        return {
            type: "points",
            earned: correct ? 1 : 0,
            total: 1,
            message: null,
        };
    }

    state: any = {
        changed: false,
    };

    componentDidMount() {
        this._isMounted = true;
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    handleChange: (arg1: React.ChangeEvent<HTMLInputElement>) => void = (e) => {
        if (!this._isMounted) {
            return;
        }

        this.setState({changed: true}, () => {
            // Wait until all components have rendered. In React 16, the
            // setState callback fires immediately after componentDidUpdate,
            // and there is no guarantee that parent/siblings components have
            // finished rendering.
            // TODO(jeff, CP-3128): Use Wonder Blocks Timing API
            // eslint-disable-next-line no-restricted-syntax
            setTimeout(() => {
                this.props.onChange(e);
                this.props.trackInteraction();
            }, 0);
        });
    };

    getUserInput: () => {
        options: any;
    } = () => {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'getOptions' does not exist on type 'ReactInstance'.
        return {options: this.refs.sortable.getOptions()};
    };

    moveOptionToIndex: (option: SortableOption, index: number) => void = (
        option,
        index,
    ) => {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'moveOptionToIndex' does not exist on type 'ReactInstance'.
        this.refs.sortable.moveOptionToIndex(option, index);
    };

    simpleValidate: (arg1: any) => any = (rubric) => {
        // If this widget hasn't been changed yet, we treat it as "empty" which
        // prevents the "Check" button from becoming active. We want the user
        // to make a change before trying to move forward. This makes an
        // assumption that the initial order isn't the correct order! However,
        // this should be rare if it happens, and interacting with the list
        // will enable the button, so they won't be locked out of progressing.
        if (!this.state.changed) {
            return {
                type: "invalid",
                message: null,
            };
        }

        return Sorter.validate(this.getUserInput(), rubric);
    };

    render(): React.ReactNode {
        const options = shuffle(
            this.props.correct,
            this.props.problemNum,
            /* ensurePermuted */ true,
        );

        const {apiOptions} = this.props;
        const marginPx = apiOptions.isMobile ? 8 : 5;

        return (
            <div className="perseus-widget-sorter perseus-clearfix">
                <Sortable
                    // @ts-expect-error - TS2322 - Type 'readonly unknown[]' is not assignable to type 'readonly string[]'.
                    options={options}
                    layout={this.props.layout}
                    margin={marginPx}
                    padding={this.props.padding}
                    onChange={this.handleChange}
                    linterContext={this.props.linterContext}
                    // eslint-disable-next-line react/no-string-refs
                    ref="sortable"
                />
            </div>
        );
    }
}

export default {
    name: "sorter",
    displayName: "Sorter",
    widget: Sorter,
    isLintable: true,
} as WidgetExports<typeof Sorter>;
