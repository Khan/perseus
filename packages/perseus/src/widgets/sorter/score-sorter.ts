import Util from "../../util";

import validateSorter from "./validate-sorter";

import type {PerseusScore} from "../../types";
import type {
    PerseusSorterScoringData,
    PerseusSorterUserInput,
} from "../../validation.types";

function scoreSorter(
    userInput: PerseusSorterUserInput,
    scoringData: PerseusSorterScoringData,
): PerseusScore {
    const validationError = validateSorter(userInput);
    if (validationError) {
        return validationError;
    }

    const correct = Util.deepEq(userInput.options, scoringData.correct);
    return {
        type: "points",
        earned: correct ? 1 : 0,
        total: 1,
        message: null,
    };
}

export default scoreSorter;
