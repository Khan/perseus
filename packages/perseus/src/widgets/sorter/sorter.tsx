import {
    type PerseusSorterWidgetOptions,
    getSorterPublicWidgetOptions,
} from "@khanacademy/perseus-core";
import {linterContextDefault} from "@khanacademy/perseus-linter";
import {scoreSorter, validateSorter} from "@khanacademy/perseus-score";
import * as React from "react";

import Sortable from "../../components/sortable";
import Util from "../../util";
import {getPromptJSON as _getPromptJSON} from "../../widget-ai-utils/sorter/sorter-ai-utils";

import type {SortableOption} from "../../components/sortable";
import type {Widget, WidgetExports, WidgetProps} from "../../types";
import type {SorterPromptJSON} from "../../widget-ai-utils/sorter/sorter-ai-utils";
import type {
    PerseusSorterRubric,
    PerseusSorterUserInput,
} from "@khanacademy/perseus-score";

const {shuffle} = Util;

type RenderProps = PerseusSorterWidgetOptions;

type Props = WidgetProps<RenderProps, PerseusSorterRubric>;

type DefaultProps = {
    correct: Props["correct"];
    layout: Props["layout"];
    padding: Props["padding"];
    problemNum: Props["problemNum"];
    onChange: Props["onChange"];
    linterContext: Props["linterContext"];
};

type State = {
    changed: boolean;
};

class Sorter extends React.Component<Props, State> implements Widget {
    _isMounted: boolean = false;

    static defaultProps: DefaultProps = {
        correct: [],
        layout: "horizontal",
        padding: true,
        problemNum: 0,
        onChange: function () {},
        linterContext: linterContextDefault,
    };

    state: State = {
        changed: false,
    };

    componentDidMount() {
        this._isMounted = true;
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    handleChange: (arg1: React.ChangeEvent<HTMLInputElement>) => void = (e) => {
        if (!this._isMounted) {
            return;
        }

        this.setState({changed: true}, () => {
            // Wait until all components have rendered. In React 16, the
            // setState callback fires immediately after componentDidUpdate,
            // and there is no guarantee that parent/siblings components have
            // finished rendering.
            // TODO(jeff, CP-3128): Use Wonder Blocks Timing API
            // eslint-disable-next-line no-restricted-syntax
            setTimeout(() => {
                this.props.onChange(e as any);
                this.props.trackInteraction();
            }, 0);
        });
    };

    getUserInput(): PerseusSorterUserInput {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'getOptions' does not exist on type 'ReactInstance'.
        const options = this.refs.sortable.getOptions();
        return {
            options,
            changed: this.state.changed,
        };
    }

    getPromptJSON(): SorterPromptJSON {
        return _getPromptJSON(this.getUserInput());
    }

    moveOptionToIndex: (option: SortableOption, index: number) => void = (
        option,
        index,
    ) => {
        // eslint-disable-next-line react/no-string-refs
        // @ts-expect-error - TS2339 - Property 'moveOptionToIndex' does not exist on type 'ReactInstance'.
        this.refs.sortable.moveOptionToIndex(option, index);
    };

    render(): React.ReactNode {
        const options = shuffle(
            this.props.correct,
            this.props.problemNum as number,
            /* ensurePermuted */ true,
        );

        const {apiOptions} = this.props;
        const marginPx = apiOptions.isMobile ? 8 : 5;

        return (
            <div className="perseus-widget-sorter perseus-clearfix">
                <Sortable
                    options={options}
                    layout={this.props.layout}
                    margin={marginPx}
                    padding={this.props.padding}
                    onChange={this.handleChange}
                    linterContext={this.props.linterContext}
                    // eslint-disable-next-line react/no-string-refs
                    ref="sortable"
                />
            </div>
        );
    }
}

export default {
    name: "sorter",
    displayName: "Sorter",
    widget: Sorter,
    isLintable: true,
    // TODO(LEMS-2656): remove TS suppression
    // @ts-expect-error: Type UserInput is not assignable to type PerseusSorterUserInput
    scorer: scoreSorter,
    // TODO(LEMS-2656): remove TS suppression
    // @ts-expect-error: Type UserInput is not assignable to type PerseusSorterUserInput
    validator: validateSorter,
    getPublicWidgetOptions: getSorterPublicWidgetOptions,
} satisfies WidgetExports<typeof Sorter>;
