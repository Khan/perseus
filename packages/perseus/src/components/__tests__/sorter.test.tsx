import * as React from "react";
import "@testing-library/jest-dom";

import {testDependencies} from "../../../../../testing/test-dependencies";
import {wait} from "../../../../../testing/wait";
import * as Dependencies from "../../dependencies";
import {renderQuestion} from "../../widgets/__tests__/renderQuestion";
import {question1} from "../__testdata__/sorter.testdata";

import type {APIOptions} from "../../types";

describe("sorter widget", () => {
    beforeEach(() => {
        /*
        Sortable misbehaves and sets state after the component has been
        unmounted. This is existing behavior and its safer to leave the existing
        implementation and swallow the warning in tests.
        */
        jest.spyOn(console, "warn").mockImplementation(() => {});
        jest.spyOn(console, "error").mockImplementation(() => {});

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue({
            ...testDependencies,
            TeX: ({
                children,
                // alias onRender to onLoad to quiet the overzealous testing-library linter
                onRender: onLoad,
            }: {
                children: React.ReactNode;
                onRender?: () => unknown;
            }) => {
                React.useLayoutEffect(() => {
                    onLoad?.();
                }, [onLoad]);
                return <span className="tex-mock">{children}</span>;
            },
        });
    });

    it("should snapshot", async () => {
        // Arrange
        jest.useRealTimers();

        const apiOptions: APIOptions = {
            isMobile: false,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);
        await wait();

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should snapshot on mobile", async () => {
        // Arrange
        jest.useRealTimers();

        const apiOptions: APIOptions = {
            isMobile: true,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);
        await wait();

        // Assert
        expect(container).toMatchSnapshot("first mobile render");
    });

    it("can be answered correctly", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);
        const sorter = renderer.findWidgets("sorter 1")[0];

        // Put the options in the correct order
        ["$0.005$ kilograms", "$15$ grams", "$55$ grams"].forEach(
            (option, index) => {
                sorter.moveOptionToIndex(option, 3);
            },
        );

        // Act
        renderer.guessAndScore();

        // assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });
    it("can be answered incorrectly", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);
        const sorter = renderer.findWidgets("sorter 1")[0];

        // Put the options in the reverse order
        ["$0.005$ kilograms", "$15$ grams", "$55$ grams"].forEach(
            (option, index) => {
                sorter.moveOptionToIndex(option, 0);
            },
        );

        // Act
        renderer.guessAndScore();

        // assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});
