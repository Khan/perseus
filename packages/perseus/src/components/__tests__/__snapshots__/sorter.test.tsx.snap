// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sorter widget should snapshot on mobile: first mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <strong>
          Arrange the following measurements in order from smallest to largest.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="perseus-widget-sorter perseus-clearfix"
          >
            <ul
              class="sortable_13d4756 perseus-sortable"
            >
              <li
                class="card_1s6cuhy-o_O-draggable_12to336-o_O-horizontalCard_9um937 perseus-interactive perseus-sortable-draggable"
                style="position: static; margin: 0px 8px 0px 0px;"
              >
                <div
                  class="perseus-renderer perseus-renderer-responsive"
                >
                  <div
                    class="paragraph"
                    data-perseus-paragraph-index="0"
                  >
                    <div
                      class="paragraph"
                    >
                      <span
                        style="white-space: nowrap;"
                      >
                        <span />
                        <span
                          class="tex-mock"
                        >
                          15
                        </span>
                        <span />
                      </span>
                       grams
                    </div>
                  </div>
                </div>
              </li>
              <li
                class="card_1s6cuhy-o_O-draggable_12to336-o_O-horizontalCard_9um937 perseus-interactive perseus-sortable-draggable"
                style="position: static; margin: 0px 8px 0px 0px;"
              >
                <div
                  class="perseus-renderer perseus-renderer-responsive"
                >
                  <div
                    class="paragraph"
                    data-perseus-paragraph-index="0"
                  >
                    <div
                      class="paragraph"
                    >
                      <span
                        style="white-space: nowrap;"
                      >
                        <span />
                        <span
                          class="tex-mock"
                        >
                          0.005
                        </span>
                        <span />
                      </span>
                       kilograms
                    </div>
                  </div>
                </div>
              </li>
              <li
                class="card_1s6cuhy-o_O-draggable_12to336-o_O-horizontalCard_9um937 perseus-interactive perseus-sortable-draggable"
                style="position: static; margin: 0px;"
              >
                <div
                  class="perseus-renderer perseus-renderer-responsive"
                >
                  <div
                    class="paragraph"
                    data-perseus-paragraph-index="0"
                  >
                    <div
                      class="paragraph"
                    >
                      <span
                        style="white-space: nowrap;"
                      >
                        <span />
                        <span
                          class="tex-mock"
                        >
                          55
                        </span>
                        <span />
                      </span>
                       grams
                    </div>
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`sorter widget should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <strong>
          Arrange the following measurements in order from smallest to largest.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="perseus-widget-sorter perseus-clearfix"
          >
            <ul
              class="sortable_13d4756 perseus-sortable"
            >
              <li
                class="card_1s6cuhy-o_O-draggable_12to336-o_O-horizontalCard_9um937 perseus-interactive perseus-sortable-draggable"
                style="position: static; margin: 0px 5px 0px 0px;"
              >
                <div
                  class="perseus-renderer perseus-renderer-responsive"
                >
                  <div
                    class="paragraph"
                    data-perseus-paragraph-index="0"
                  >
                    <div
                      class="paragraph"
                    >
                      <span
                        style="white-space: nowrap;"
                      >
                        <span />
                        <span
                          class="tex-mock"
                        >
                          15
                        </span>
                        <span />
                      </span>
                       grams
                    </div>
                  </div>
                </div>
              </li>
              <li
                class="card_1s6cuhy-o_O-draggable_12to336-o_O-horizontalCard_9um937 perseus-interactive perseus-sortable-draggable"
                style="position: static; margin: 0px 5px 0px 0px;"
              >
                <div
                  class="perseus-renderer perseus-renderer-responsive"
                >
                  <div
                    class="paragraph"
                    data-perseus-paragraph-index="0"
                  >
                    <div
                      class="paragraph"
                    >
                      <span
                        style="white-space: nowrap;"
                      >
                        <span />
                        <span
                          class="tex-mock"
                        >
                          0.005
                        </span>
                        <span />
                      </span>
                       kilograms
                    </div>
                  </div>
                </div>
              </li>
              <li
                class="card_1s6cuhy-o_O-draggable_12to336-o_O-horizontalCard_9um937 perseus-interactive perseus-sortable-draggable"
                style="position: static; margin: 0px;"
              >
                <div
                  class="perseus-renderer perseus-renderer-responsive"
                >
                  <div
                    class="paragraph"
                    data-perseus-paragraph-index="0"
                  >
                    <div
                      class="paragraph"
                    >
                      <span
                        style="white-space: nowrap;"
                      >
                        <span />
                        <span
                          class="tex-mock"
                        >
                          55
                        </span>
                        <span />
                      </span>
                       grams
                    </div>
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
